server/etcdserver/txn/txn_test.go

```diff
@@ -94,3 +98,289 @@ func TestWriteTxnPanic(t *testing.T) {
 
 	assert.Panics(t, func() { Txn(ctx, zaptest.NewLogger(t), txn, false, s, &lease.FakeLessor{}) }, "Expected panic in Txn with writes")
 }
+
+func TestCheckTxnAuth(t *testing.T) {
```

Comment:
Last thing, the test function is just too long. Could you:
* Move auth setup to dedicated function
* Move variables like `inRangeCompare` out of the test.

For example create function
```
func setupAuth(t *testing.T, be backend.Backend) {
  simpleTokenTTLDefault := 300 * time.Second
	tokenTypeSimple := "simple"
	dummyIndexWaiter := func(index uint64) <-chan struct{} {
		ch := make(chan struct{}, 1)
		go func() {
			ch <- struct{}{}
		}()
		return ch
	}

	tp, _ := auth.NewTokenProvider(zaptest.NewLogger(t), tokenTypeSimple, dummyIndexWaiter, simpleTokenTTLDefault)

	as := auth.NewAuthStore(lg, schema.NewAuthBackend(lg, be), tp, 4)

	// create "root" user and "foo" user with limited range
	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "root"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.RoleAdd(&pb.AuthRoleAddRequest{Name: "rw"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{
		Name: "rw",
		Perm: &authpb.Permission{
			PermType: authpb.READWRITE,
			Key:      []byte("foo"),
			RangeEnd: []byte("zoo"),
		},
	}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "root", Password: "foo"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserAdd(&pb.AuthUserAddRequest{Name: "foo", Password: "foo"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "root", Role: "root"}); err != nil {
		t.Fatal(err)
	}
	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "foo", Role: "rw"}); err != nil {
		t.Fatal(err)
	}

	if err := as.AuthEnable(); err != nil {
			t.Fatal(err)
	}
}

In the end test function should look like 
```
func TestCheckTxnAuth(t *testing.T) {
	lg := zaptest.NewLogger(t)

	be, _ := betesting.NewDefaultTmpBackend(t)
	defer betesting.Close(t, be)
	setupAuth(t, be)
		tests := []struct {
		name       string
		txnRequest *pb.TxnRequest
		err        error
	}{
		{
			name: "Out of range compare is unauthorized",
			txnRequest: &pb.TxnRequest{
				Compare: []*pb.Compare{outOfRangeCompare},
			},
			err: auth.ErrPermissionDenied,
		},
	...
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := CheckTxnAuth(as, &auth.AuthInfo{Username: "foo", Revision: 8}, tt.txnRequest)
			if err != tt.err {
				t.Errorf("expected error to be: %v; got: %v", tt.err, err)
			}
		})
	}
}

}
```

Reply:
Perfect!